File: services/internal/domain/track.go
package domain

import (
	"strings"
	"time"

	"github.com/google/uuid"
)

type Track struct {
	ID    uuid.UUID
	JobID uuid.UUID

	// suno_audio_id — id конкретного результата у Suno (важно для идемпотентности callback).
	SunoAudioID string

	Title string
	Tags  *string

	DurationSec float64

	// Координаты объекта в MinIO/S3.
	// Это то, что хранится в БД и не протухает.
	AudioBucket string
	AudioKey    string

	// Обложка (если есть)
	ImageBucket *string
	ImageKey    *string

	// Временные/вычисляемые поля для отдачи клиенту.
	// В БД их хранить не нужно, т.к. presigned URL протухают.
	AudioURL string

	StreamURL *string
	ImageURL  *string

	IsFavorite bool

	CreatedAt time.Time
}

func (t Track) Validate() error {
	if t.ID == uuid.Nil {
		return InvalidInput(ErrTrackIDRequired)
	}
	if t.JobID == uuid.Nil {
		return InvalidInput(ErrTrackJobIDRequired)
	}
	if strings.TrimSpace(t.SunoAudioID) == "" {
		return InvalidInput(ErrTrackSunoAudioIDReq)
	}
	if strings.TrimSpace(t.Title) == "" {
		return InvalidInput(ErrTrackTitleRequired)
	}

	// Главное изменение: для сохранения в БД требуем bucket/key,
	// а не presigned URL.
	if strings.TrimSpace(t.AudioBucket) == "" || strings.TrimSpace(t.AudioKey) == "" {
		return InvalidInput(ErrTrackAudioStorageRequired)
	}

	return nil
}

// ValidateForResponse — строгая валидация для того, что отдаём клиенту.
// (Например, перед маппингом в DTO.)
func (t Track) ValidateForResponse() error {
	if err := t.Validate(); err != nil {
		return err
	}
	if strings.TrimSpace(t.AudioURL) == "" {
		return InvalidInput(ErrTrackAudioURLRequired)
	}
	return nil
}

File: apps/mobile_flutter/lib/shared/domain/track.dart
class Track {
  final String id;
  final String jobId;
  final String sunoAudioId;
  final String title;
  final String? tags;
  final double durationSec;
  final String audioUrl;
  final String? streamUrl;
  final String? imageUrl;
  final bool isFavorite;
  final DateTime createdAt;

  const Track({
    required this.id,
    required this.jobId,
    required this.sunoAudioId,
    required this.title,
    this.tags,
    required this.durationSec,
    required this.audioUrl,
    this.streamUrl,
    this.imageUrl,
    required this.isFavorite,
    required this.createdAt,
  });


  factory Track.fromJson(Map<String, dynamic> json) {
    return Track(
      id: json['id'] as String,

      jobId: json['job_id'] as String,

      sunoAudioId: json['suno_audio_id'] as String,

      title: json['title'] as String,

      tags: json['tags'] as String?,

      durationSec: (json['duration_sec'] as num).toDouble(),

      audioUrl: json['audio_url'] as String,

      streamUrl: json['stream_url'] as String?,

      imageUrl: json['image_url'] as String?,

      isFavorite: json['is_favorite'] as bool,

      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'job_id': jobId,
      'suno_audio_id': sunoAudioId,
      'title': title,
      'tags': tags,
      'duration_sec': durationSec,
      'audio_url': audioUrl,
      'stream_url': streamUrl,
      'image_url': imageUrl,
      'is_favorite': isFavorite,
      'created_at': createdAt.toIso8601String(),
    };
  }
}